File: py/benchmark.py
import numpy as np
from radon import radon
from time import perf_counter

# Radon transform parameters
n_rows = 512
n_cols = 512
n_rays = 200
n_slopes = 200

# Load input image
image = np.loadtxt("../inputs/lena.csv", delimiter = ',')

# Benchmark loop
n_trials = 10
times = [0.] * n_trials
for idx in range(n_trials):
    t_start = perf_counter()
    radon_transform = radon(image, n_rows, n_cols, n_rays, n_slopes)
    t_stop = perf_counter() 
    times[idx] = t_stop - t_start

# Saving times
np.savetxt("../outputs/py_times.csv", times, delimiter = ",")

# Printing statistics
print(f"mean: {np.mean(times)*1e3:.2f} ms, var: {np.var(times)*1e3:.2f}")

File: src/benchmark.rs
use crate::array::{mean, var};
use crate::csv_rw::{read_csv, write_vec_csv};
use std::time::Instant;

mod array;
mod csv_rw;
mod radon;

fn main() {
    // Radon transform parameters
    let n_rows = 512_usize;
    let n_cols = 512_usize;
    let n_rays = 200_usize;
    let n_slopes = 200_usize;

    // Load input image
    let input_path = "inputs/lena.csv".to_string();
    let image = read_csv(input_path).expect("Error reading input.");

    // Benchmark loop
    let n_trials = 10;
    let mut times: Vec<f32> = vec![];
    for _ in 0..n_trials {
        let t_start = Instant::now();
        radon::radon(&image, n_rows, n_cols, n_rays, n_slopes);
        let t_stop = t_start.elapsed();
        times.push(t_stop.as_secs_f32());
    }

    // Saving times
    write_vec_csv(&times, "outputs/rs_times.csv".to_string()).expect("Couldn't save times.");

    // Printing statistics
    println!(
        "mean: {:.2} ms, var: {:.2}",
        mean(&times) * 1000.,
        var(&times) * 1000.
    );
}
